### PR TITLE
Translator refactor

### PR DESCRIPTION
**Pull request by vince62s**
Goal is to:
- eliminate all the option list of the Translator Class and pass the full opt/model_opt
- factor the decoder / generator to be used by:
   1) translate_batch (now using same dimension order as fat ie batch x beam for tensors)
   2) _fast_translate_batch
   3) Gold scores retrieval (_score_target)
- fix Gold Scores in copy_attn mode

Ready for review.
Tested in text mode for copy / non-copy
Tests for Image / audio are welcome.

### CODE REVIEW

**Comment by sebastianGehrmann** [comment]
Do you have more information on the bug in non-fast?

**Comment by vince62s** [comment]
if you look at the PR, here https://github.com/OpenNMT/OpenNMT-py/pull/1043/files#diff-b228765b5cdb90ddc4bf121dcdad5eabR322
I copy-pasted more or less the code from the fast version, but obvioulsy
I must have missed something in the tensors from the non-fast version here:
https://github.com/OpenNMT/OpenNMT-py/pull/1043/files#diff-b228765b5cdb90ddc4bf121dcdad5eabL644
Guillaume changed the behavior of collapse_copy_scores to handle the batched version of decoding, but maybe dimensions are switched.

**Comment by guillaumekln** [comment]
I did not change the behavior of `collapse_copy_scores`, I just [extended it](https://github.com/OpenNMT/OpenNMT-py/commit/49185121e46c4f65d68101590bad231a8dd73e4f#diff-eb5f4f6d050a86f260354d8f92901af1) to support variable batch size and switched beam/batch dimensions. What is the exact bug you are referring to? I tested again on master and got the same results in fast and default modes so I'm not sure what I missed.

**Comment by vince62s** [comment]
Yes I know, this is what I missed, now I need to use the same dimension order, so in fact batch_dim=0 for both.
Still have an issue on results, digging more.

**Comment by vince62s** [comment]
ok, this is not optimal (since non-fast and fast beam search have their dimensions reversed) but it's working for inference in non copy_attn as well copy_attn.
Now we need to modify the _score_target code to output gold scores based on the new -decode_and_generate()

File: onmt/translate/beam.py [hunk 1]
```diff
@@ -189,11 +189,11 @@ class GNMTGlobalScorer(object):
        beta (float):  coverage parameter
     """
 
-    def __init__(self, alpha, beta, cov_penalty, length_penalty):
-        self.alpha = alpha
-        self.beta = beta
-        penalty_builder = penalties.PenaltyBuilder(cov_penalty,
-                                                   length_penalty)
+    def __init__(self, opt):
```
Comment (guillaumekln) [inline onmt/translate/beam.py]: Not sure that's an improvement. It makes sense for high level classes to get the `opt` argument but other than that it is better to keep fully defined arguments IMO.

File: onmt/translate/translator.py [hunk 2]
```diff
@@ -343,6 +316,55 @@ def _run_encoder(self, batch, data_type):
...
+            # returns [(batch_size x beam_size) , vocab ] when 1 step
+            # or [ tgt_len, batch_size, vocab ] when full sentence
+        else:
+            attn = dec_attn["copy"]
+            scores = self.model.generator(dec_out.view(-1, dec_out.size(2)),
+                                          attn.view(-1, attn.size(2)),
+                                          src_map)
+            # here we have scores [tgt_lenxbatch, vocab] or [beamxbatch, vocab]
+            if batch_offset is None:
+                scores = scores.view(batch.batch_size, -1, scores.size(-1))
```
Comment (guillaumekln) [inline onmt/translate/translator.py]: This line does the same thing as the `else` clause but is less general. It can be removed.

Comment (guillaumekln) [reply onmt/translate/translator.py]: My comment was invalid.

**Comment by vince62s** [comment]
Guillaume, since I changed the dimensions order for the non-fast, you noticed I am calling collapse_copy_score with batch_dim=0 in all cases.
We may then change the way collapse_copy_score is called in training and subsequently remove / simplify this part of the code.

**Pull request merged**